### PR TITLE
eth/downloader: fix dysfunctional ignore list hidden by generic set

### PR DESCRIPTION
**Pull request by karalabe**
This PR fixes an ugly issue hidden by the usage of a generic set implementation, which failed to catch mismatching types being placed inside it.

In essence, we were using a generic `fatih.Set` to accumulate a list of hashes that a node didn't have in order to prevent re-requesting them multiple times. The original code to populate this set iterated over a `map[hash]Header` construct, placing all the selected hashes into said set. After rewriting this construct to simply `[]Header`, the for each construct (`for hash, _ := range headers`) remained valid, but [instead of  adding hashes, it added indexes](https://github.com/ethereum/go-ethereum/pull/1960/files#diff-bdc55f77f09f52de6a3820c6619900caL880), essentially rendering the ignore list dysfunctional. The end result is that the same peer may be tried multiple times for the same block it already demonstrated not to have.

This fix also puts a hard cap on the number of hashes that can be ignored from a particular peer, randomly dropping old ones when that limit is exceeded.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Nov 4 10:26:27 UTC 2015_

File: eth/downloader/peer.go [hunk 1]
```diff
@@ -305,13 +311,42 @@ func (p *peer) Demote() {
...
+	p.ignoreLock.Lock()
+	defer p.ignoreLock.Unlock()
+
+	p.ignored[hash] = struct{}{}
+	for len(p.ignored) > maxIgnoredHashes {
+		for drop, _ := range p.ignored {
+			delete(p.ignored, drop)
+			break
+		}
+	}
```
Comment (fjl) [inline eth/downloader/peer.go]: Maybe drop before adding, otherwise you might drop the one you just added.

Comment (karalabe) [inline eth/downloader/peer.go]: Yeah, thought about that but the probability is so small (1/4097) that I figured it's clearer this way logically. Though if you think that way's better, I'll fix it up.

Comment (fjl) [inline eth/downloader/peer.go]: Could also maintain an LRU list alongside the map.

Comment (karalabe) [inline eth/downloader/peer.go]: True, but I'm not sure it's worth the complexity of adding another construct in there. It's true that this map is cimple and stupid, but as long as it works I'd rather not complicate matters.

Comment (karalabe) [inline eth/downloader/peer.go]: Fixed the ordering issue.

Comment (fjl) [inline eth/downloader/peer.go]: This is not a blocking issue. The map prevents making pointless requests *again*, so it is already a second-level mechanism. If possible, we need to try and not make pointless requests in the first place.

Comment (karalabe) [inline eth/downloader/peer.go]: Taking into consideration remote nodes's TDs to figure out what to request is definitely something I do want to incorporate, but that's a tad larger fix so it's going to be done in a different PR. This one is just meant to hotfix the issue.

File: eth/downloader/peer.go [hunk 2]
```diff
@@ -305,13 +311,42 @@ func (p *peer) Demote() {
...
+	p.ignored[hash] = struct{}{}
+}
+
+// Ignored retrieves whether the hash of a blockchain item is on the peers ignore
+// list or not (i.e. whether we know that the peer does not have it).
```
Comment (fjl) [inline eth/downloader/peer.go]: The fact that we need to explain what the 'ignore list' is makes me wonder whether this is the right name.
Consider calling it `failedReq` (and the map `failCache`) or `lacks` (and the map `lacking`).
It should also not be exported.

Comment (karalabe) [inline eth/downloader/peer.go]: My export style is that I make public all methods that should be called outside of the struct itself, and keep private methods private to the struct itself too. It makes code much readable knowing when you're calling API methods and when you're mucking around with internals.

Comment (fjl) [inline eth/downloader/peer.go]: Hm, can't argue with personal style. I just don't like stuff that looks like it is public but isn't.
Any idea about the name?

**Comment by fjl** [comment]
:+1:

**Comment by karalabe** [comment]
I've updated the names to:

 * `ignored` -> `lacking`
 * `Ignore` -> `MarkLacking`
 * `Ignored` -> `Lacks`

**Comment by fjl** [comment]
`lackingLock.Lock` looks funny, but I do like the method names more than the 'Ignore`-based names.
